src/dwarf_parser.cpp

```diff
@@ -0,0 +1,171 @@
+#include "dwarf_parser.h"
+
+#ifdef HAVE_LIBDW
+
+#include "log.h"
+#include "utils.h"
+#include <dwarf.h>
+#include <elfutils/libdw.h>
+
+#ifdef HAVE_BCC_ELF_FOREACH_SYM
+#include <bcc/bcc_elf.h>
+#include <bcc/bcc_syms.h>
+#endif
+
+namespace bpftrace {
+
+struct FuncInfo
+{
+  std::string name;
+  Dwarf_Die die;
+};
+
+Dwarf::Dwarf(const std::string &file_path) : file_path(file_path)
+{
+  Dwfl_Callbacks callbacks = {
+    .find_debuginfo = dwfl_standard_find_debuginfo,
+    .section_address = dwfl_offline_section_address
+  };
+  dwfl = dwfl_begin(&callbacks);
+  if (dwfl_report_offline(dwfl, file_path.c_str(), file_path.c_str(), -1) ==
+      NULL)
+  {
+    LOG(WARNING) << "Could not read DWARF info from " + file_path;
+  }
+  dwfl_report_end(dwfl, NULL, NULL);
+
+  Dwarf_Addr bias;
+  has_data_ = dwfl_nextcu(dwfl, NULL, &bias) != NULL;
```

Comment:
Nit incoming:

As an alternative to `has_data()`, we can make the `Dwarf` constructor private and have a `friend` factory method return a `std::optional<Dwarf>`. If the factory method returns an instance, then the callers may assume it contains data. This has a nice property where `Dwarf` methods can assume the instance contains data and skip `has_data()` checks.

This alternative has its own drawbacks in that the class contract is a bit different. And `BTF` class also does the `has_data()` thing. So I don't mind either way -- just something to consider.

Reply:
I like this since it is more transparent for users of the `Dwarf` class.

But, in order to manage libdw objects correctly, we don't want to allow copying `Dwarf` instances around. So, I prefer to return `std::unique_ptr<Dwarf>` instead of `std::optional<Dwarf>`.